src/ivc/incrementally_verifiable_computation.rs

```diff
@@ -228,17 +231,89 @@ where
 
     pub fn verify<const T: usize, RP1, RP2>(
         &mut self,
-        _pp: &PublicParams<'_, A1, A2, T, C1, C2, SC1, SC2, RP1, RP2>,
-        _steps_count: usize,
-        _z0_primary: [C1::Scalar; A1],
-        _z0_secondary: [C2::Scalar; A2],
+        pp: &PublicParams<'_, A1, A2, T, C1, C2, SC1, SC2, RP1, RP2>,
+        num_steps: usize,
+        z0_primary: [C1::Scalar; A1],
+        z0_secondary: [C2::Scalar; A2],
     ) -> Result<(), Error>
     where
-        RP1: ROPair<C1::Scalar>,
-        RP2: ROPair<C2::Scalar>,
+        RP1: ROPair<C1::Scalar, Config = MainGateConfig<T>>,
+        RP2: ROPair<C2::Scalar, Config = MainGateConfig<T>>,
     {
-        // TODO #31
-        todo!("Logic at `RecursiveSNARK::verify`")
+        let is_step_zero = num_steps == 0;
+        let is_num_step_not_match = num_steps != self.step;
+        let is_inputs_match = self.primary.z_0 != z0_primary || self.secondary.z_0 != z0_secondary;
+        if is_step_zero || is_num_step_not_match || is_inputs_match {
+            return Err(Error::IVCParamNotMatch);
+        }
```

Comment:
It's a terrible design in the Nova codebase, please let's not repeat it. Each of these bugs is a separate one with its own type and context

Reply:
And I'm not sure we should be accepting and validating these arguments at all. Can you give a scenario when it's needed exactly like this for this class? 

When verify & fold are separate entities it's understandable, but here I was thinking of consolidating the final design into a single `IVC::fold(... num_steps ...)` method that does everything itself, including verify

Reply:
I think the checking of `self.step=num_steps` is not very important. Just for the case that we want to run a given number of steps and do the final verification.

It's not a good idea to consolidating.  When the prover is honest, the verification (it's costly) at each step should always pass, if we combine them, it will slow down the prover speed.  
The `fn verify` is only called at the end after folding `num_steps` by verifier. And in future, we will replace part by another zkSNARK

Reply:
I.e., for a specific this type we don't need anything at all but one such function, we even risk more when we take `pp` by reference every time and have to recalculate their `digest`

Reply:
`fn fold` for internally for dev purpose is totally fine. 
And to avoid recalculate `digest`, we can put the calculated hash into IVC:
`IVC.pp_hash: once_cell`